// rclnodejs/generated/rqt_py_common/rqt_py_common__msg__ArrayVal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ValWrapper = require('../../generated/rqt_py_common/rqt_py_common__msg__Val.js');
const ArrayValRefStruct = StructType({
  vals: ArrayType(ValWrapper.refObjectType, 5),
});
const ArrayValRefArray = ArrayType(ArrayValRefStruct);
const ArrayValRefStructArray = StructType({
  data: ArrayValRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ArrayValWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ArrayValRefStruct(other._refObject.toObject());
      this._wrapperFields.vals = ValWrapper.createArray();
      this._wrapperFields.vals.copy(other._wrapperFields.vals);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ArrayValRefStruct();
    this._wrapperFields.vals = ValWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ArrayValWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ArrayValArrayWrapper;
  }
  static get ArrayType() {
    return ArrayValArrayWrapper;
  }
  static get refObjectArrayType() {
    return ArrayValRefStructArray
  }
  static get refObjectType() {
    return ArrayValRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    for (let i = 0; i < this._wrapperFields.vals.data.length; i++) {
      this._refObject.vals[i] = this._wrapperFields.vals.data[i].freeze(own, checkConsistency);
      this._refObject.vals[i] = this._wrapperFields.vals.data[i].refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.vals = refObject.vals;
    this._wrapperFields.vals.size = 5
    for (let i = 0; i < 5; i++) {
      this._wrapperFields.vals.data[i].copyRefObject(refObject.vals[i]);
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    for (let i = 0; i < 5; i++) {
      ValWrapper.freeStruct(refObject.vals[i]);
    }
  }
  static destoryRawROS(msg) {
    ArrayValWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rqt_py_common', subFolder: 'msg', interfaceName: 'ArrayVal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get vals() {
    return this._wrapperFields.vals;
  }
  set vals(value) {
    if (value.length !== 5) {
      throw new RangeError('The length of the array must be 5.');
    }
    this._wrapperFields.vals.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ArrayValRefStruct(refObject.toObject());
    this._refObject.vals = refObject.vals;
    this._wrapperFields.vals.size = 5
    for (let i = 0; i < 5; i++) {
      this._wrapperFields.vals.data[i].copyRefObject(refObject.vals[i]);
    }
  }
  copy(other) {
    this._refObject = new ArrayValRefStruct(other._refObject.toObject());
    this._wrapperFields.vals.copy(other._wrapperFields.vals);
  }
  static get classType() {
    return ArrayValWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"vals","type":{"isArray":true,"arraySize":5,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":"rqt_py_common","type":"Val","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rqt_py_common","type":"ArrayVal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ArrayVal"};
  }
  hasMember(name) {
    let memberNames = ["vals"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ArrayValArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ArrayValWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ArrayValWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ArrayValArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ArrayValArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ArrayValRefArray(size);
    this._refObject = new ArrayValRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ArrayValWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ArrayValArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ArrayValArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ArrayValWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ArrayValWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ArrayValArrayWrapper;
  }
}
module.exports = ArrayValWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "vals",
      "type": {
        "isArray": true,
        "arraySize": 5,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": "rqt_py_common",
        "type": "Val",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rqt_py_common",
    "type": "ArrayVal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ArrayVal"
}
*/
