// rclnodejs/generated/rcl_interfaces/rcl_interfaces__msg__IntraProcessMessage.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const IntraProcessMessageRefStruct = StructType({
  publisher_id: primitiveTypes.uint64,
  message_sequence: primitiveTypes.uint64,
});
const IntraProcessMessageRefArray = ArrayType(IntraProcessMessageRefStruct);
const IntraProcessMessageRefStructArray = StructType({
  data: IntraProcessMessageRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class IntraProcessMessageWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new IntraProcessMessageRefStruct(other._refObject.toObject());
      this._publisher_idIntialized = true;
      this._message_sequenceIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new IntraProcessMessageRefStruct();
    this._publisher_idIntialized = false;
    this._message_sequenceIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new IntraProcessMessageWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new IntraProcessMessageArrayWrapper;
  }
  static get ArrayType() {
    return IntraProcessMessageArrayWrapper;
  }
  static get refObjectArrayType() {
    return IntraProcessMessageRefStructArray
  }
  static get refObjectType() {
    return IntraProcessMessageRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._publisher_idIntialized) {
        throw new TypeError('Invalid argument: publisher_id in IntraProcessMessage');
      }
      if (!this._message_sequenceIntialized) {
        throw new TypeError('Invalid argument: message_sequence in IntraProcessMessage');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._publisher_idIntialized = true;
    this._refObject.publisher_id = refObject.publisher_id;
    this._message_sequenceIntialized = true;
    this._refObject.message_sequence = refObject.message_sequence;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    IntraProcessMessageWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'msg', interfaceName: 'IntraProcessMessage'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get publisher_id() {
    if (!this._publisher_idIntialized) {
      return undefined;
    }
    return this._refObject.publisher_id;
  }
  set publisher_id(value) {
    this._publisher_idIntialized = true;
    this._refObject.publisher_id = value;
  }
  get message_sequence() {
    if (!this._message_sequenceIntialized) {
      return undefined;
    }
    return this._refObject.message_sequence;
  }
  set message_sequence(value) {
    this._message_sequenceIntialized = true;
    this._refObject.message_sequence = value;
  }
  copyRefObject(refObject) {
    this._refObject = new IntraProcessMessageRefStruct(refObject.toObject());
    this._publisher_idIntialized = true;
    this._message_sequenceIntialized = true;
  }
  copy(other) {
    this._refObject = new IntraProcessMessageRefStruct(other._refObject.toObject());
    this._publisher_idIntialized = true;
    this._message_sequenceIntialized = true;
  }
  static get classType() {
    return IntraProcessMessageWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"publisher_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"message_sequence","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"IntraProcessMessage","stringUpperBound":null,"isPrimitiveType":false},"msgName":"IntraProcessMessage"};
  }
  hasMember(name) {
    let memberNames = ["publisher_id","message_sequence"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class IntraProcessMessageArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof IntraProcessMessageWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new IntraProcessMessageWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IntraProcessMessageArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IntraProcessMessageArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new IntraProcessMessageRefArray(size);
    this._refObject = new IntraProcessMessageRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new IntraProcessMessageWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof IntraProcessMessageArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "IntraProcessMessageArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      IntraProcessMessageWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return IntraProcessMessageWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return IntraProcessMessageArrayWrapper;
  }
}
module.exports = IntraProcessMessageWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "publisher_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "message_sequence",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rcl_interfaces",
    "type": "IntraProcessMessage",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "IntraProcessMessage"
}
*/
